File: .claude/.planning/lang-spike-document-module/fsharp/Document.fs
/// Document — agregado + máquina de estados.
/// Espelha `financial/domain/document/{types,document,errors}.ts`.
module Financial.Domain.Document

open Financial.Domain.Money

/// DocumentStatus — DU (espelha o string-literal-union + `const _: never` do TS).
/// Um `match` incompleto gera warning FS0025 → com `<WarningsAsErrors>FS0025</WarningsAsErrors>`
/// no `.fsproj` vira ERRO de compilação (ligar isso é obrigatório para igualar o Rust).
type DocumentStatus =
    | Draft
    | Open
    | Approved
    | Transmitted
    | Refused
    | Paid
    | PartiallyReconciled
    | Reconciled

/// Erros do agregado — DU (espelha o string-literal-union de erros do TS).
type DocumentError =
    | NotDraft
    | NetValueRequired

/// Código EN kebab-case para a borda HTTP (espelha o dicionário de erros do projeto).
let errorCode =
    function
    | NotDraft -> "document-not-draft"
    | NetValueRequired -> "document-net-value-required"

/// Recorte fiel do agregado (subset — o real tem ~20 campos em `DocumentCore`). Record = imutável.
type Document =
    { Id: string
      DocumentNumber: string
      Status: DocumentStatus
      GrossValue: Money
      NetValue: Money option // `netValue: Money | null` → option nativo
      Version: int64 }

/// Submete Draft → Open. Imutável: `{ doc with ... }` é o que o `{ ...prev, status }` do TS imita.
/// Note: este match usa apenas os construtores (sem guard `when`), então a exaustividade é checada.
let submit (doc: Document) : Result<Document, DocumentError> =
    match doc.Status, doc.NetValue with
    | Draft, Some _ -> Ok { doc with Status = Open; Version = doc.Version + 1L }
    | Draft, None -> Error NetValueRequired
    | _ -> Error NotDraft

File: .claude/.planning/lang-spike-document-module/rust/document.rs
//! Document — agregado + máquina de estados.
//! Espelha `financial/domain/document/{types,document,errors}.ts`.

use crate::money::Money;

/// DocumentStatus — DU nativa (espelha o string-literal-union + `const _: never` do TS).
/// `match` sobre isto é exaustivo POR PADRÃO: adicionar variante quebra a compilação em todo `match`.
/// Sem flag de projeto (o F# precisa de `FS0025-as-error`; Rust não).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DocumentStatus {
    Draft,
    Open,
    Approved,
    Transmitted,
    Refused,
    Paid,
    PartiallyReconciled,
    Reconciled,
}

/// Erros do agregado — enum (espelha o string-literal-union de erros do TS).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DocumentError {
    NotDraft,
    NetValueRequired,
}

impl DocumentError {
    /// Código EN kebab-case para a borda HTTP (espelha o dicionário de erros do projeto).
    pub fn code(self) -> &'static str {
        match self {
            DocumentError::NotDraft => "document-not-draft",
            DocumentError::NetValueRequired => "document-net-value-required",
        }
    }
}

/// Recorte fiel do agregado (subset — o real tem ~20 campos em `DocumentCore`).
#[derive(Debug, Clone)]
pub struct Document {
    pub id: String, // DocumentId (newtype no real)
    pub document_number: String,
    pub status: DocumentStatus,
    pub gross_value: Money,
    pub net_value: Option<Money>, // `netValue: Money | null` → Option nativo
    pub version: i64,
}

impl Document {
    /// Submete Draft → Open (espelha a transição `submit` do agregado).
    /// Imutável: retorna uma CÓPIA nova via struct update (equivale ao `{ ...prev, status: 'Open' }`).
    pub fn submit(&self) -> Result<Document, DocumentError> {
        if self.status != DocumentStatus::Draft {
            return Err(DocumentError::NotDraft);
        }
        if self.net_value.is_none() {
            return Err(DocumentError::NetValueRequired);
        }
        Ok(Document {
            status: DocumentStatus::Open,
            version: self.version + 1,
            ..self.clone()
        })
    }
}
